doc: adding a note about config expressions (#25218)
keycloak/keycloak: docs/guides/server/configuration.adoc

. Use the `[-cf|--config-file]` command line parameter to include that file.

Similarly, you can also store Quarkus properties in a Java KeyStore.

Note that some Quarkus properties are already mapped in the {project_name} configuration, such as `quarkus.http.port` and similar essential properties. If the property is used by Keycloak, defining that property key in `quarkus.properties` has no effect. The Keycloak configuration value takes precedence over the Quarkus property value.

== Values

Keycloak depends upon Quarkus / MicroProfile for processing configuration values. Be aware this means that value expression are supported, for example value=$\{some_value}$\{some_other_value}, which will evaluate to something other than the literal value.

> To disable expression evaluation, the \ character is used as an escape character where needed. In particular it must be used to escape the usage of $ characters when they appear to define an expression or are repeated.  For example instead of a configuration value of my$$password, use my\$\$password.

Reviewer: Some comments here around quoting contents for the docs (apply here if necessary):

* If you struggle with escaping things for AsciiDoc, consider putting it into double-plus signs for pass-through content (`++passthrough++`).
* If you struggle with escaping freemarker stuff (like things in curly braces), wrap it as no-parse (`<#noparse>...</#noparse>`)

> To disable expression evaluation, the \ character is used as an escape character where needed. In particular it must be used to escape the usage of $ characters when they appear to define an expression or are repeated.  For example instead of a configuration value of my$$password, use my\$\$password.

Author: > Seems triple \$ is needed?

No, the key here is using \\ to escape the $.  Using \\$\\$\\$ would evaluate to 3 dollar signs, not 2.  There may be a fix upstream to correct the $$ behavior, so we don't want to document relying on that - https://github.com/smallrye/smallrye-config/issues/1056

> To disable expression evaluation, the \ character is used as an escape character where needed. In particular it must be used to escape the usage of $ characters when they appear to define an expression or are repeated.  For example instead of a configuration value of my$$password, use my\$\$password.

Reviewer 2: @shawkins Not sure why (haven't investigated) but if I use something like `--db-username posgre\$\$\$ql`, it evaluates to `posgre$$ql` (looking at DB log of failed auth attempt).

> To disable expression evaluation, the \ character is used as an escape character where needed. In particular it must be used to escape the usage of $ characters when they appear to define an expression or are repeated.  For example instead of a configuration value of my$$password, use my\$\$password.

Author: @vmuzikar please review again, I've incorporated most the suggestions and further clarified shell handling.

> To disable expression evaluation, the \ character is used as an escape character where needed. In particular it must be used to escape the usage of $ characters when they appear to define an expression or are repeated.  For example instead of a configuration value of my$$password, use my\$\$password.

Reviewer 2: I looked into it more. The issue with 3 `$$$` was truly caused by the shell backslash escaping as you hinted.

However, I think the docs change is a little bit confusing still. We're mixing together two things: the SmallRye expression escaping, and shell expressions escaping. For example, let's assume the `my$$password` value. When specified in shell, it needs to be entered as `"my\$\$password"` to mitigate for shell expressions. However, when specified via a properties file it needs to be entered as `my$$$$password` to mitigate SmallRye expressions (which are not evaluated on other config sources AFAIK). This is currently not captured in the docs.

> To disable expression evaluation, the \ character is used as an escape character where needed. In particular it must be used to escape the usage of $ characters when they appear to define an expression or are repeated.  For example instead of a configuration value of my$$password, use my\$\$password.

Author: > We're mixing together two things: the SmallRye expression escaping, and shell expressions escaping.

That's not intended - shell escaping only comes into this at the very end as an additional concern on top of the escaping needed for Microprofile config / smallrye.

> For example, let's assume the my$$password value. When specified in shell, it needs to be entered as "my\$\$password" to mitigate for shell expressions.

That's not at all what I'm trying to convey - it doesn't matter where the value appears, the $$ needs to be escaped.

> However, when specified via a properties file it needs to be entered as my$$$$password to mitigate SmallRye expressions

That's not at all what I would recommend.  As far as I can find, that's undocumented behavior - unless I'm missing something it doesn't appear in Microprofile config nor quarkus docs.  I do see references to \ as the escape character - and it's explicitly handled in the smallrye code as such wrt $.

> which are not evaluated on other config sources AFAIK

The removal of $$ is certainly happening when processing cli args, are you suggesting that you don't expect nested expressions to be resolved with cli or env variables?

. Use the `[-cf|--config-file]` command line parameter to include that file.

Similarly, you can also store Quarkus properties in a Java KeyStore.

Note that some Quarkus properties are already mapped in the {project_name} configuration, such as `quarkus.http.port` and similar essential properties. If the property is used by Keycloak, defining that property key in `quarkus.properties` has no effect. The Keycloak configuration value takes precedence over the Quarkus property value.

=== Using special characters in values

> {project_name} depends upon Quarkus and MicroProfile for processing configuration values. Be aware that value expressions are supported - for example `$\{some_key}`, will evaluate to the value of `some_key`.

Reviewer 3: ```suggestion
{project_name} depends upon Quarkus and MicroProfile for processing configuration values. Be aware that value expressions are supported. For example, `$\{some_key}` evaluates to the value of `some_key`.
```

> To disable expression evaluation, the `\` character is used as an escape character. In particular it must be used to escape the usage of `$` characters when they appear to define an expression or are repeated. For example if you want the configuration value `my$$password`, use `my\$\$password` instead.  Note that the `\` character requires additional escaping or quoting when using most unix shells, or when it appears in properties files.

Reviewer 3: ```suggestion
To disable expression evaluation, the `\` character functions as an escape character. In particular, it must be used to escape the usage of `$` characters when they appear to define an expression or are repeated. For example, if you want the configuration value `my$$password`, use `my\$\$password` instead.  Note that the `\` character requires additional escaping or quoting when using most unix shells, or when it appears in properties files.
```

> To disable expression evaluation, the `\` character is used as an escape character. In particular it must be used to escape the usage of `$` characters when they appear to define a expression or are repeated. For example if you want the configuration value `my$$password`, use `my\$\$password` instead.  Note that the `\` character requires additional escaping or quoting when using most unix shells, or when it appears in properties files.  
> For example bash single quotes will preseve the single backslash --db-password=`'my\$\$password'`, or with bash double quotes you will need an additional backslash --db-password=`"my\\$\\$password"`. Similarly in a properties file backslash characters must themselves be escaped: kc.db-password=`my\\$\\$password`

Reviewer 2: Thanks for the update! I think it's clearer now.

One last nitpick. I think we should be putting whole key+value in the backticks. Seems It's more common throughout the docs.
```suggestion
To disable expression evaluation, the `\` character is used as an escape character. In particular it must be used to escape the usage of `$` characters when they appear to define a expression or are repeated. For example if you want the configuration value `my$$password`, use `my\$\$password` instead.  Note that the `\` character requires additional escaping or quoting when using most unix shells, or when it appears in properties files.  
For example bash single quotes will preserve the single backslash `--db-password='my\$\$password'`, or with bash double quotes you will need an additional backslash `--db-password="my\\$\\$password"`. Similarly in a properties file backslash characters must themselves be escaped: `kc.db-password=my\\$\\$password`
```

> For example bash single quotes will preserve the single backslash `--db-password='my\$\$password'`, or with bash double quotes you will need an additional backslash `--db-password="my\\$\\$password"`. Similarly in a properties file backslash characters must themselves be escaped: `kc.db-password=my\\$\\$password`

Reviewer 3: ```suggestion
For example, bash single quotes preserve the single backslash `--db-password='my\$\$password'`. Also, with bash double quotes, you need an additional backslash `--db-password="my\\$\\$password"`. Similarly in a properties file, backslash characters must also be escaped: `kc.db-password=my\\$\\$password`
```
